dmoj/cptbox/isolate.py

```diff
@@ -201,176 +202,290 @@ def __init__(self, read_fs, write_fs=None, writable=(1, 2)):
+    def handle_open(self, syscall: str, file_reg: int = 0, flag_reg: int = 1) -> HandlerCallback:
+        return self.wrap_access_check(
+            syscall,
+            self.main_access_check(
+                syscall, self._fs_jail_getter_from_open_flags_reg(flag_reg), self._dirfd_getter_cwd, file_reg
+            ),
+        )
 
-    def check_file_access_at(self, syscall, kind, file_reg=1) -> HandlerCallback:
-        fs_jail = self._fs_jail(attrgetter('uarg2'), kind)
+    def handle_openat(self, syscall: str, dir_reg: int = 0, file_reg: int = 1, flag_reg: int = 2) -> HandlerCallback:
+        return self.wrap_access_check(
+            syscall,
+            self.main_access_check(
+                syscall,
+                self._fs_jail_getter_from_open_flags_reg(flag_reg),
+                self._dirfd_getter_from_reg(dir_reg),
+                file_reg,
+            ),
+        )
 
+    def handle_fstat(self, syscall: str, dir_reg: int = 0, file_reg: int = 1) -> HandlerCallback:
+        @self.wrap_access_check_dec(syscall)
         def check(debugger: Debugger) -> bool:
-            file, error = self.read_path(syscall, debugger, getattr(debugger, 'uarg%d' % file_reg))
-            if error is not None:
-                return error
+            rel_file = self.get_rel_file(debugger, file_reg)
 
             # FIXME(tbrindus): defined here because FreeBSD 13 does not
             # implement AT_EMPTY_PATH, and 14 is not yet released (but does).
             AT_EMPTY_PATH = 0x1000
             # FIXME(tbrindus): we always follow symlinks, regardless of whether
             # AT_SYMLINK_NOFOLLOW is set. This may result in us denying files
             # we otherwise wouldn't have.
-            if kind == FilesystemSyscallKind.FSTAT and file == '' and debugger.uarg3 & AT_EMPTY_PATH:
+            if rel_file == '' and debugger.uarg3 & AT_EMPTY_PATH:
                 # If pathname is an empty string, operate on the file referred to
                 # by dirfd (which may have been obtained using the open(2) O_PATH
                 # flag). In this case, dirfd can refer to any type of file, not
                 # just a directory, and the behavior of fstatat() is similar to
                 # that of fstat(). If dirfd is AT_FDCWD, the call operates on the
                 # current working directory.
+                # We already allowed this one way or another, don't check again
                 return True
 
-            file, error = self._file_access_check(file, debugger, fs_jail(debugger), dirfd=debugger.arg0)
-            if not error:
-                return True
+            dirfd = self.get_reg_value(debugger, dir_reg)
+            full_path = self.get_full_path(debugger, rel_file, dirfd)
+            return self._raw_access_check(syscall, debugger, full_path, self.read_fs_jail)
+
+        return check
+
+    @staticmethod
+    def wrap_access_check_dec(syscall: str) -> Callable[[AccessChecker], HandlerCallback]:
```

Comment:
I'm not sure what `check_dec` is supposed to mean, I've read the code & tried to think, but failed

Reply:
Supposed to be a decorator, but removed entirely after this new push.